### ISSUE TITLE
Usage of Matrix Factorization Trainer for Recommendation 

### ISSUE THREAD

**Issue by asthana86** [dotnet/machinelearning#1806]
When using Matrix Factorization Trainer:

 ```csharp
var trainer = mlcontext.Recommendation().Trainers.MatrixFactorization
                                                     ("userIdEncoded", "movieIdEncoded", "rating"));
```
When using other trainers:

 ```csharp
var trainer = mlContext.Regression.Trainers.StochasticDualCoordinateAscent
                                                       (label: "Label", features: "Features");
``` 

Is the difference in usage prop vs. method by design? Also there is a difference in the order of parameters being passed. First parameter is Label vs. Features being used.

**Comment by wschin** [comment]
```csharp
var trainer = mlcontext.Recommendation().Trainers.MatrixFactorization
                                                    ("userIdEncoded", "movieIdEncoded", "rating"));
```
is equivalent to
```csharp
var trainer = mlcontext.Recommendation().Trainers.MatrixFactorization
                                                    (matrixColumnIndexColumnName: "userIdEncoded", matrixRowIndexColumnName: "movieIdEncoded", labelColumn: "rating"));
```
according to its signature
```csharp
        /// <summary>
        /// Initializing a new instance of <see cref="MatrixFactorizationTrainer"/>.
        /// </summary>
        /// <param name="env">The private instance of <see cref="IHostEnvironment"/>.</param>
        /// <param name="matrixColumnIndexColumnName">The name of the column hosting the matrix's column IDs.</param>
        /// <param name="matrixRowIndexColumnName">The name of the column hosting the matrix's row IDs.</param>
        /// <param name="labelColumn">The name of the label column.</param>
        /// <param name="advancedSettings">A delegate to apply all the advanced arguments to the algorithm.</param>
        public MatrixFactorizationTrainer(IHostEnvironment env,
            string matrixColumnIndexColumnName,
            string matrixRowIndexColumnName,
            string labelColumn = DefaultColumnNames.Label,
            Action<Arguments> advancedSettings = null)
            : base(env, LoadNameValue)
```

Instead of a single feature `Column` and a label `Column`, matrix factorization requires row index `Column`, column index `Column`, and label `Column`. For example, assume that
```
row index column = [0, 3, 1]
column index column = [2, 1, 0]
label column = [7, 7, 8]
```
the 4-by-3 rating matrix being factorized may be
```
[? ? 7]
|8 ? ? |
|? ? ? |
[? 7 ?]
```
where `?` happens at row `u` and column `v` means user `u` never rates `v` in your training data. Note that I assume those IDs are 0-based indexes.

**Comment by asthana86** [comment]
Two questions still: 

Why are these different:
            var foo = mlContext.Regression.Trainers;
            var foo_bar = mlContext.BinaryClassification.Trainers;
            var foo_moo_bar = mlContext.Recommendation().Trainers;

vs 
           var foo_moo_bar = mlContext.Recommendation.Trainers

and then the order of parameters again for consistency? other trainers seem to take Label as the first parameter for input. so this instead. 

var trainer = mlcontext.Recommendation().Trainers.MatrixFactorization
                                                    (labelColumn: "rating", matrixColumnIndexColumnName: "userIdEncoded", matrixRowIndexColumnName: "movieIdEncoded"));

**Comment by wschin** [comment]
@assafi, for your second question, matrix factorization is a special problem. It's not standard regression/classification which maps a feature vector to a label, so it looks not very bad to have label as the last argument. Of course, you can submit a PR if this doesn't look good enough for you. It should be a minor change.

**Comment by singlis** [comment]
Hi @asthana86, 

For the question about the differences in the mlContext api calls, please  see issue #1770 as this addresses your question. It is because Recommendation is not part of the nuget core package and therefore is defined as an extension rather than a property of MLContext. 

For the ordering of label columns in the matrix factorization, we can create an issue on this or use this as an the issue for tracking.

**Comment by asthana86** [comment]
I am not sure if #1770 addresses my issue. #1770 is about discoverability, the fix we came up for that was one needs to acquire the MatrixFactorization Nuget for now. I do agree with @GalOshri its not the best experience but even with that nuget acquisition the usage pattern should remain the same.  

Given the need to acquire an additional NuGet is there no way to have the API be consistent like the one that follows? It just looks a bit odd from a user perspective. 

mlcontext.Recommendation.Trainers.MatrixFactorization

instead of

mlcontext.Recommendation().Trainers.MatrixFactorization

In terms of the other issue, it might be worth it creating a list of all learners on order of input parameters /output parameters and follow it across. I have been working on exporting these samples to 0.8 and after playing with Regression, Classification the recommendation ML task is a bit less consistent to other MLTasks.

**Comment by asthana86** [comment]
Sounds great. Thanks for capturing this.

**Comment by singlis** [comment]
Thanks @asthana86.  For reference, I filed the two issues
#1827 The trainer API for MLContext is inconsistent for learner in external nugets
#1826 - MatrixFactorization construction parameters are not consistent with other learners.

I will keep this issue open for a few days in case there is any other things to discuss.

**Comment by singlis** [comment]
Closing as we have the other two issues filed.
